File: backend/src/main/java/com/netcompany/coe/login/data/ThrottlingDatabase.java
package com.netcompany.coe.login.data;

import com.google.common.collect.ImmutableList;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Stream;

import static java.util.Collections.emptyList;

@Repository
public class ThrottlingDatabase {

    private static final int TIME_SPAN_MINUTES = 1;
    private static final int MAX_ALLOWED_ATTEMPTS_IN_TIME_SPAN = 5;

    private final Map<String, List<LocalDateTime>> userLoginAttempts = new ConcurrentHashMap<>();

    public boolean isIllegalLoginAttempt(String username) {
        return loginAttemptsInTimeSpan(username)
                .count() > MAX_ALLOWED_ATTEMPTS_IN_TIME_SPAN;
    }

    public void recordLoginAttempt(String username) {
        userLoginAttempts.put(
                username,
                ImmutableList.<LocalDateTime>builder()
                        .addAll(loginAttemptsInTimeSpan(username)
                                .iterator())
                        .add(LocalDateTime.now())
                        .build());
    }

    public void successfulLogin(String username) {
        userLoginAttempts.remove(username);
    }

    private Stream<LocalDateTime> loginAttemptsInTimeSpan(String username) {
        return userLoginAttempts.getOrDefault(username, emptyList()).stream()
                .filter(ThrottlingDatabase::isInTimeSpan);
    }

    private static boolean isInTimeSpan(LocalDateTime timestamp) {
        return timestamp.isAfter(LocalDateTime.now().minusMinutes(TIME_SPAN_MINUTES));
    }
}

File: backend/src/main/java/com/netcompany/coe/login/data/ThrottlingDatabase.kt
package com.netcompany.coe.login.data

import org.springframework.stereotype.Repository
import java.time.LocalDateTime
import java.util.concurrent.ConcurrentHashMap

private const val TIME_SPAN_MINUTES = 1L
private const val MAX_ALLOWED_ATTEMPTS_IN_TIME_SPAN = 5

@Repository
class ThrottlingDatabase {

    private val userLoginAttempts = ConcurrentHashMap<String, List<LocalDateTime>>()

    fun isIllegalLoginAttempt(username: String): Boolean {
        return loginAttemptsInTimeSpan(username).count() > MAX_ALLOWED_ATTEMPTS_IN_TIME_SPAN
    }

    fun recordLoginAttempt(username: String) {
        userLoginAttempts[username] = loginAttemptsInTimeSpan(username).plus(LocalDateTime.now())
    }

    fun successfulLogin(username: String) {
        userLoginAttempts.remove(username)
    }

    private fun loginAttemptsInTimeSpan(username: String): List<LocalDateTime> {
        return userLoginAttempts[username]?.filter(::isInTimeSpan) ?: listOf()
    }

}

private fun isInTimeSpan(timestamp: LocalDateTime): Boolean {
    return timestamp.isAfter(LocalDateTime.now().minusMinutes(TIME_SPAN_MINUTES))
}
